<h1>Assignment 1</h1>

<p>In this first assignment we are focusing on a classification problem.<br>
Our goal will be to use both the <b>K-Nearest Neighbors</b> and the <b>Decision Tree</b> techniques and try to compare the predictive models obtained.</p>

<h3>The Dataset</h3>
<p>For this assignment we will be using the sonar_train.csv data set</p>
<p>This data set contains signals obtained from a variety of different aspect angles, spanning 90 degrees for mines and 180 degrees for rocks. Each sample is a set of 60 numbers in the range 0.0 to 1.0, where each number represents the energy within a particular frequency band (integrated over a certain period of time). The output target contains the letter R if the object is a rock and M if it is a mine (metal cylinder).</p>

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
%matplotlib inline

In [2]:
#Opening the sonar_train.csv file
data = pd.read_csv('sonar.csv',',')
#data = pd.read_csv('sonar.csv')
###Display informations and the different column/features names
data.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 208 entries, 0 to 207
Data columns (total 61 columns):
Band1     208 non-null float64
Band2     208 non-null float64
Band3     208 non-null float64
Band4     208 non-null float64
Band5     208 non-null float64
Band6     208 non-null float64
Band7     208 non-null float64
Band8     208 non-null float64
Band9     208 non-null float64
Band10    208 non-null float64
Band11    208 non-null float64
Band12    208 non-null float64
Band13    208 non-null float64
Band14    208 non-null float64
Band15    208 non-null float64
Band16    208 non-null float64
Band17    208 non-null float64
Band18    208 non-null float64
Band19    208 non-null float64
Band20    208 non-null float64
Band21    208 non-null float64
Band22    208 non-null float64
Band23    208 non-null float64
Band24    208 non-null float64
Band25    208 non-null float64
Band26    208 non-null float64
Band27    208 non-null float64
Band28    208 non-null float64
Band29    208 non-null floa

<p>In this data set each band is a feature and the target is the type of the object</p>
<h4> <u>QUESTION 1 :</u> How many samples and how many features do we have?</h4>
<p><i> We have 208 samples and 61 features</i></p>

In [3]:
###using the train_test_split command split the data set in two : 'train' and 'test' with a ratio ~75%-25%
from sklearn.cross_validation import train_test_split
x=data.drop(['Type'],axis=1)
y=data['Type']
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=0)


In [4]:
#The last column in our sets correspond to the target
###Count how many of 'R' and 'M' you have in total
print 'The number of R is',sum(y=='R')
print 'The number of M is',sum(y=='M')

The number of R is 97
The number of M is 111


In [5]:
#In order to make targets more easily usable we will replace 'R' and 'M' by 0 and 1 respectively.
###Do it for both sets using the command : .replace()

yy_train=y_train.replace('R',0).replace('M',1)
yy_test=y_test.replace('R',0).replace('M',1)


<h3>K-Nearest Neighbors approach</h3>

<p>In this section we will apply the <b>K-Nearest Neighbors</b> technique to the data set</p>

In [7]:
from sklearn.neighbors import KNeighborsClassifier
###Fille the brackets to use the brute force approach and 7 neighbors
knn = KNeighborsClassifier(algorithm='auto',n_neighbors=7)

### Train the Nearest Neighbor Classifier "knn" on the training set
knn.fit(x_train,yy_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=7, p=2,
           weights='uniform')

<h4> <u>QUESTION 2 :</u> Which kind of impact has the choice of the algorithm?</h4>
<p><i>The computer time used to carry out the classification depends on the choice of algorithm </i></p>

In [8]:
#At any time it is possible to recall the parameters using :
knn.get_params()

{'algorithm': 'auto',
 'leaf_size': 30,
 'metric': 'minkowski',
 'metric_params': None,
 'n_jobs': 1,
 'n_neighbors': 7,
 'p': 2,
 'weights': 'uniform'}

<p>We now want to test the accuracy of the nearest neighbor classifier</p>


In [9]:
###Compute the target prediction for the elements of the test set and compare with their true value

result_test=knn.predict(x_test)
result_test==yy_test

12     False
80     False
33     False
5      False
187     True
83     False
116     True
122     True
90      True
154     True
45     False
156     True
52     False
189     True
96      True
86      True
204     True
37      True
66      True
18      True
170    False
15      True
7      False
55     False
92     False
134     True
125     True
124     True
158     True
184     True
75      True
149    False
138    False
71      True
186     True
145    False
176     True
118     True
16      True
135     True
190     True
22      True
104     True
141     True
4      False
74     False
136     True
44     False
63      True
108     True
26     False
101     True
Name: Type, Length: 52, dtype: bool

In [10]:
#scikit actually includes an accuracy evaluation function :
###Compute the accuracy using knn.score()

len(yy_test)==len(result_test)
knn.score(x_test,yy_test)

0.65384615384615385

<h4> <u>QUESTION 3 :</u> Is the accuracy acceptable?</h4>
<p><i>No</i></p>

<p>In the following cell repeat the training and the test steps with different knn parameters and try to find a good combination.</p>

In [11]:
knn = KNeighborsClassifier(algorithm='auto',n_neighbors=4)
### Train the Nearest Neighbor Classifier "knn" on the training set
knn.fit(x_train,yy_train)
knn.score(x_test,yy_test)




0.82692307692307687

<h4> <u>QUESTION 4 :</u> What is the best accuracy you get? What can you conclude?</h4>
<p><i>My best accuracy is 0.83. Since the data is short, we can conclude based on this result</i></p>

<h3>Decision Tree</h3>

<p>Using the same sets we now investigate the efficiency of a <b>decision tree classifier<b> </p>

In [18]:
from sklearn import tree
###
dtr1 = tree.DecisionTreeClassifier()

###train the decision tree classifier "dtr"
dtr1.fit(x_train,yy_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            presort=False, random_state=None, splitter='best')

In [19]:
#Same as previously we can display some informations
dtr1.get_params()

{'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': None,
 'max_leaf_nodes': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'presort': False,
 'random_state': None,
 'splitter': 'best'}

<h4> <u>QUESTION 5 :</u> From the list above, give the keywords influencing the test performed at a decison node?</h4>
<p><i>Criterion,min_samples_leaf,min_samples_split and splitter</i></p>

<p>In the follwing, using the work done previously test the accuracy of the decision tree classifier.</p>

In [20]:
dtr1.score(x_test,yy_test)






0.82692307692307687

<p>Hereafter, using different parameters for our decision tree, try to improve the accuracy</p>

<h4> <u>QUESTION 6 :</u> Explain which parameter(s) you plan to change and why?</h4>
<p><i>I plan to change the max_depth </i></p>


In [22]:
dtr = tree.DecisionTreeClassifier(max_depth=56)
dtr.fit(x_train,yy_train)
dtr.score(x_test,yy_test)






0.78846153846153844

<p>The results are different but obviously when we identify mines we don't want to be wrong.</p>

<h4> <u>QUESTION 7 :</u> Which one of the two methods is better? Is the accuracy acceptable? Why?</h4>
<p><i>The Decision Tree is the best method. The accuracy is acceptable since the accuracy is high compared with the number of dataset that we have</i></p>

<h4> <u>QUESTION 8 :</u> What can you conclude from the whole study</h4>
<p><i>I could conclude that the decision tree method works better with this data. Also, changing some parameter in both method increases the accuracy.</i></p>


<h4> <u>BONUS :</u> Copy/Paste the code to generate a pdf file containing the best tree and attach the file generated to your assignment when sending everything to me</h4>

In [25]:
import pydot
from sklearn.externals.six import StringIO
from sklearn import tree

dot_data = StringIO()
f = tree.export_graphviz(dtr,out_file=dot_data,
                         feature_names=x.columns,
                         class_names='Type',
                         filled=True,rounded=True)
graph = pydot.graph_from_dot_data(dot_data.getvalue())
graph.write_pdf('1619_Akinlabi Emmanuel_A1.pdf')



True